tests/models/shadow/commit.p

```diff
@@ -0,0 +1,115 @@
+//
+// The Commit Protocol takes as input a set of produce requests and commits the
+// associated batches. This involves (1) building and uploading the batch data
+// to cloud storage, (2) committing metadata about the object into each
+// partition targetted by the produce requests and (3) acknowledging each client
+// with the final committed offset.
+//
+
+// An L0d object stores payload data from produced batches without offsets.
+// In the model, the position of an entry in an L0d object is used as its
+// identifer in placeholder batches. The values in the L0d object are unique
+// batch identifiers generated by the client. They represent the "content" of
+// the batch, and must not be used by the model except for validation.
+type L0d_object = seq[int];
+
+machine CommitProtocol {
+  var storage: Storage;
+  var requests: seq[produce_request];
+
+  // the L0d object that will be built and uploaded
+  var object: L0d_object;
+  var object_id: int;
+
+  // mapping from offset in object to source request
+  var object_to_request: seq[int];
+  // mapping from append request to offset of data in object
+  var append_to_object: map[int, int];
+
+  // holds the next request id
+  var request_id: int;
+
+  start state Init {
+    entry (input: (storage: Storage, requests: seq[produce_request])) {
+      storage = input.storage;
+      requests = input.requests;
+      goto Commit;
+    }
+  }
+
+  state Commit {
+    entry {
+      build_L0d();
+      upload_L0d();
+      append_placeholders();
+    }
+
+    // Respond to the produce request with the committed offset.
+    on append_response_event do (response: append_response) {
+      var offset: int;
+      var request: produce_request;
+
+      // map the partition append response back the original produce request
+      offset = object_to_request[append_to_object[response.request_id]];
```

Comment:
_for the reader (me) this is the most confusing usage of `offset` of all. On this line it means one thing and in the same event handler on line 58 now it means a different thing_